### PR TITLE
Cleanup reputation API + back with an EnumMap instead of array

### LINKED ISSUE #7142: Can't reset villager reputation
### Expected behavior

Reputation gets reset

### Observed/Actual behavior

I cant reset the villager's reputation with the api

### Steps/models to reproduce
`rep.setReputation(ReputationType.MINOR_NEGATIVE, 0);`
`rep.setReputation(ReputationType.MAJOR_NEGATIVE, 0);`
`villager.setReputation(p.getUniqueId(), rep);`

### Plugin and Datapack List

Irrelevant

### Paper version

This server is running Paper version git-Paper-66 (MC: 1.18) (Implementing API version 1.18-R0.1-SNAPSHOT) (Git: dce79f3)
You are running the latest version

### Other

https://github.com/PaperMC/Paper/blob/master/patches/server/0422-Add-villager-reputation-API.patch
lines 54-58 should be the problem

A possible workaround:

`Map<UUID, Reputation> reps = villager.getReputations();`
`Reputation rep = reps.get(p.getUniqueId());`
`if (rep != null) {`
`	rep.setReputation(ReputationType.MINOR_NEGATIVE, 0);`
`	rep.setReputation(ReputationType.MAJOR_NEGATIVE, 0);`
`}`
`villager.clearReputations();`
`villager.setReputations(reps);`

### PR DESCRIPTION
**Pull request by Owen1212055**
Fixes https://github.com/PaperMC/Paper/issues/7142

### CODE REVIEW

**Review by kennytv — CHANGES_REQUESTED** [review]
needs another default return value for unfilled entries to not update any unset reputation type

**Comment by Owen1212055** [comment]
Now uses an enum map, where it will only remove values if explicitly set to 0.

**Review by kennytv — APPROVED** [review]

**Review by SpiritOTHawk — APPROVED** [review]

File: patches/server/0411-Add-villager-reputation-API.patch [hunk 1]
```diff
@@ -32,37 +17,54 @@ index 7a7c92f1a026116958ad24312df358a703834369..0de65462956fa734b6405614e0471616
...
-+            int val;
-+            if ((val = rep.getReputation(com.destroystokyo.paper.entity.villager.ReputationType.MAJOR_NEGATIVE)) != 0) this.entries.put(GossipType.MAJOR_NEGATIVE, val);
-+            if ((val = rep.getReputation(com.destroystokyo.paper.entity.villager.ReputationType.MAJOR_POSITIVE)) != 0) this.entries.put(GossipType.MAJOR_POSITIVE, val);
-+            if ((val = rep.getReputation(com.destroystokyo.paper.entity.villager.ReputationType.MINOR_NEGATIVE)) != 0) this.entries.put(GossipType.MINOR_NEGATIVE, val);
-+            if ((val = rep.getReputation(com.destroystokyo.paper.entity.villager.ReputationType.MINOR_POSITIVE)) != 0) this.entries.put(GossipType.MINOR_POSITIVE, val);
-+            if ((val = rep.getReputation(com.destroystokyo.paper.entity.villager.ReputationType.TRADING)) != 0) this.entries.put(GossipType.TRADING, val);
++            for (GossipType type : TYPES) {
++                com.destroystokyo.paper.entity.villager.ReputationType api = toApi(type);
++
++                int reputation = rep.getReputation(api);
```
Comment (kennytv) [inline patches/server/0411-Add-villager-reputation-API.patch]: Move this into the if

**Comment by Owen1212055** [comment]
Resolved + Rebased + yes

**Comment by Owen1212055** [comment]
Rebased

**Review by lynxplay — APPROVED** [review]

**Pull request merged**